rest/admin_api.go

```diff
@@ -1423,78 +1424,123 @@ func (h *handler) handleDeleteDB() error {
-	if redact {
-		if salt == "" {
-			salt = uuid.New().String()
+	xattrsObject := []byte(base.EmptyDocument)
+
+	// inject xattrs as top-level properties in the doc - this isn't *quite* representative of the actual stored doc,
+	// since they are outside the doc body as separate objects, but we'll stamp a marker value in each xattr object to tell us whether it was injected or inline in the original doc body.
+	for _, k := range xattrKeys {
+		v := xattrValues[k]
+		if !redact {
+			xattrsObject, err = base.InjectJSONProperties(xattrsObject, base.KVPair{Key: k, Val: json.RawMessage(v)})
+			if err != nil {
+				return base.HTTPErrorf(http.StatusInternalServerError, "couldn't inject xattr %q into response: %s", k, err)
+			}
+		} else {
+			// xattr-specific redaction
+			switch k {
+			case base.SyncXattrName:
+				redactedV, err := db.RedactRawSyncData(v, redactSalt)
+				if err != nil {
+					return base.HTTPErrorf(http.StatusInternalServerError, "couldn't redact sync data: %s", err)
+				}
+				xattrsObject, err = base.InjectJSONProperties(xattrsObject, base.KVPair{Key: k, Val: json.RawMessage(redactedV)})
+				if err != nil {
+					return base.HTTPErrorf(http.StatusInternalServerError, "couldn't inject sync data into response: %s", err)
+				}
+			case userXattrKey:
+				// include the key but not the value so we can see _something_ was present.
+				xattrsObject, err = base.InjectJSONPropertiesFromBytes(xattrsObject, base.KVPairBytes{Key: k, Val: []byte(`"redacted"`)})
+				if err != nil {
+					return base.HTTPErrorf(http.StatusInternalServerError, "couldn't inject user xattr into response: %s", err)
+				}
+			}
 		}
-		syncData = doc.SyncData.HashRedact(salt)
 	}
 
-	rawBytes, err = base.InjectJSONProperties(rawBytes, base.KVPair{Key: base.SyncPropertyName, Val: syncData})
+	responseBody, err = base.InjectJSONProperties(responseBody, base.KVPair{Key: "_xattrs", Val: json.RawMessage(xattrsObject)})
 	if err != nil {
-
-		return err
+		return base.HTTPErrorf(http.StatusInternalServerError, "couldn't inject xattrs into response: %s", err)
 	}
 
-	if h.db.Options.UserXattrKey != "" {
-		metaMap, err := doc.GetMetaMap(h.db.Options.UserXattrKey)
-		if err != nil {
-			return err
-		}
-
-		rawBytes, err = base.InjectJSONProperties(rawBytes, base.KVPair{Key: "_meta", Val: metaMap})
-		if err != nil {
-			return err
-		}
-	}
 	base.Audit(h.ctx(), base.AuditIDDocumentMetadataRead, base.AuditFields{
-		base.AuditFieldDocID: docid,
+		base.AuditFieldDocID: docID,
 	})
 	if includeDoc {
+		// try to extract a revision from the sync data
+		docCurrentRev := "unknown"
+		if syncData, ok := xattrValues[base.SyncXattrName]; ok {
+			var sdRev struct {
+				Rev channels.RevAndVersion `json:"rev"`
+			}
+			if err := json.Unmarshal(syncData, &sdRev); err != nil {
+				return base.HTTPErrorf(http.StatusInternalServerError, "couldn't unmarshal rev from SyncData for doc %q: %s", base.UD(docID), err)
+			}
+			docCurrentRev = sdRev.Rev.RevTreeID
```

Comment:
This is almost `db.revOnlySyncData`. I'm fine with this, but I had to look up why this was not using `CurrentRev` using `RevTreeID` (becuase it's not sync data).

Reply:
Yeah, the persisted version of SyncData looks a bit different than the Unmarshalled version which is why I had to redefine some things.

There's not many places (if anywhere) we interact with sync data directly as it is represented marshalled.